Replace the current color picker with Coloris JS

KTS915 opened the pull request:
## Description
This PR proposes replacing the current color picker — a combination of `wp-admin/js/color-picker.js` and `wp-admin/js/iris.min.js` with Coloris JS: https://coloris.js.org/

## Motivation and context
This is part of the continuing quest to remove usage of jQuery UI in core. The current picker uses `wp-admin/js/iris.min.js`, which has dependencies of both `jquery-ui-draggable` and `jquery-ui-slider`, so it needs to be replaced, preferably by something which uses only vanilla JS and has no dependencies. The replacement must also be accessible.

Coloris JS meets all these requirements and (in my opinion, at least) looks good. It has a list of strings used, which are also made available for translation/localization. These are, by default:
```
open: 'Open color picker',
close: 'Close color picker',
clear: 'Clear the selected color',
marker: 'Saturation: {s}. Brightness: {v}.',
hueSlider: 'Hue slider',
alphaSlider: 'Opacity slider',
input: 'Color value field',
format: 'Color format',
swatch: 'Color swatch',
instruction: 'Saturation and brightness selector. Use up, down, left and right arrow keys to select.'
```
In addition, the `wp-admin/js/coloris.min.js` file is only 3.5kB in size, compared to 23.6kB for `wp-admin/js/iris.min.js` (which itself also relies on `wp-admin/js/color-picker.js`, `jquery-ui-draggable`, `jquery-ui-slider`, `jquery-ui-core` and, of course, jQuery itself).

## How has this been tested?
This has been tested on my localhost installation. It needs extensive testing and is proposed for CP v2.1 rather than 2.0.

It's also possible that others might prefer an alternative library. I tried a couple but Coloris seemed to suit us best because it's very lightweight, accessible, and it was very simple to integrate because it just requires the addition of the data attribute `data-coloris` to an `input` field with `type="text"`.

## Screenshot After
![Screenshot at 2024-01-04 17-07-45](https://github.com/ClassicPress/ClassicPress/assets/4127662/1b629e90-5915-4dbf-8033-210a66ecc8a3)

## Types of changes
- New feature

This is not a breaking change since everything should continue to work fine. The files currently used for the color picker will not be removed from the core bundle, so any theme or plugin that calls them directly will still be able to use them.

xxsimoxx commented:
<img width="309" alt="image" src="https://github.com/ClassicPress/ClassicPress/assets/29772709/55a80c19-c04e-4f2c-9f0b-812c68f8509e">

The color picker is not showing up in my local installation (MAMP, SCRIPT_DEBUG=true, Chrome).

I see that there are no unminified files. Don't know if we need those and then let the build process minify.

mattyrob commented:
IMO, the way this is currently implemented would be a breaking change. Any plugin or theme relying on the presence of `wp-color-picker` would no longer work once this is implemented.

The updated JS and CSS needs to be added using the same script name to avoid breaking. And I imagine that the hard coded strings will need some attention as well to ensure these can be translated.

KTS915 commented:
@xxsimoxx We shouldn't need unminified files. The PRS with SortableJS all use its minified file. So there's something else going on there. My guess it's a `z-index` issue. I had to bump it up to `999999` from the default to get it to show up on my install; maybe it needs to go even higher.

@mattyrob Good points, as always! I'll give some further thought to this.

KTS915 commented:
I think I've now addressed all the issues raised.

xxsimoxx commented:
Here it's not working. Tested with CP TwentySeventeen and Generate Press.

<img width="299" alt="image" src="https://github.com/ClassicPress/ClassicPress/assets/29772709/1ba96342-9975-4d8f-a779-0bff0cec38ee">

KTS915 commented:
@xxsimoxx Thanks! I tried it on TwentyFifteen and it worked, so I presumed it would work on other themes. But trying on TwentySeventeen now, I can reproduce the issue. I will play with it some more.

mattyrob commented on src/wp-admin/js/customize-controls.js:
```diff
@@ -4151,57 +4155,84 @@
 				updating = false,
-				picker;
+				picker,
+				__ = wp.i18n.__;
```
ClassicPress currently does not include the [i18n package](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-i18n/), so we need to consider if we begin to bundle it, or revert to using `wp_localize_script()`.

KTS915 replied:
`__ = wp.i18n.__` is already used in `wp-admin/js/color-picker.js`. In fact, I just did a quick `grep` of the `wp-admin/js` folder, and found it's used in all the following files:
```
user-profile.js
site-health.js
password-strength-meter.js
postbox.js
post.js
privacy-tools.js
edit-comments.js
common.js
theme-plugin-editor.js
updates.js
image-edit.js
color-picker.js:
```
In `wp-includes/js` there is also the `media-views.js` file.

So I think we need to include that package or else we're going to have a lot of files to re-write.

KTS915 commented:
I have now addressed the issues raised by @xxsimoxx with the TwentySeventeen theme.

However, I have found that there's a problem with themes (like TwentyFifteen and TwentySixteen) that provide an option for changing the Base Color Scheme. The code I have written works when the user first changes the Scheme. But if they try to change it again, nothing happens.

The console reports: `Uncaught Error: cannot call methods on wpColorPicker prior to initialization; attempted to call method 'defaultColor'`

Setting individual colors is unaffected by this.

Refreshing the page makes it possible to change the Base again, but again only one time before the same error occurs. So far, everything I've tried to re-initialize wpColorPicker results also in Iris being used as the color picker instead of Coloris.

What this does illustrate is how hacky the Customizer code really is. So much (especially of jQuery UI modules)  is constantly being destroyed and refreshed because it's full of components that don't really go together. By contrast, none of the PRs I have submitted to replace jQuery UI involves destroying or re-initializing anything because it's all designed to work together.

KTS915 commented:
I've now realized that the reason for the issue I remarked upon above with the Color Base Scheme is that, at present, it's still using Iris and not Coloris (because Coloris is designed to work on `input type="text"` fields and not `select` fields. So I need to find a way to make Coloris work with a `select` field, or — perhaps more likely — come up with an alternative.

xxsimoxx commented:
Something different but it's not working.
And console error for missing `src/wp-admin/css/color-picker.min.css`.


<img width="302" alt="image" src="https://github.com/ClassicPress/ClassicPress/assets/29772709/1652073c-bb19-4d9a-9b28-f6ea6903a4ff">

KTS915 commented:
I am closing this because I can't see a way to achieve what needs to be done using Coloris without either breaking changes or some very hacky code. But I have a new PR that takes a different approach that I think will get the job done.
